workshop1/slides/src/slides.md

```diff
@@ -0,0 +1,664 @@
+  { name :: String
+  , age  :: Int
+  } deriving Show
+```
+
+* `Person` is the type being defined
+* `MakePerson` is the constructor function
+* `name, age` are fields with associated types
+* `deriving Show` allows the use of the `show` function
+  * `show` converts structure to a string
+
+## Working with Data Types
+
+* Creating a new data type:
+
+```haskell
+MakePerson :: String -> Int -> Person
+```
+* Accessing the data type:
+
+```haskell
+name :: Person -> String
+age :: Person -> Int
+```
+![](images/simple-datatype.png)
+
+## The Maybe Type
+
+* What happens if we cannot parse a candidate?
+* Need some way to represent an invalid row:
+
+```haskell
+data Maybe a = Nothing | Just a
+```
+
+* `Maybe` is the type being defined
+* `a` is a *type variable* (could be any type)
+* `Nothing, Just` are *both* constructors
+  * `Nothing :: Maybe a`
+  * `Just :: a -> Maybe a`
+
+## The List Type
+
+* How do we represent a list of candidates?
+* Use singly-linked lists:
+  * `[]` is the empty list
+  * `:` lets you append values to the front
+  * `[a,b,c]` == `(a : b : c : [])`
+
+![](images/list.png){ height=40% }
+
+## The Data Model
+
+* Define our basic candidate type
+
+```haskell
+data Candidate = Candidate
+  { ...
+  } deriving Show
+```
+
+## The Data Model
+
+* Define our basic candidate type
+
+```haskell
+data Candidate = Candidate
+  { constituency :: String
+  , party        :: String
+  , sex          :: String
+  , averageAge   :: Double
+  } deriving Show
+```
+
+## The Data Model
+
+* Define our basic candidate type
+
+```haskell
+data Candidate = Candidate
+  { constituency :: String
+  , party        :: String
+  , sex          :: String
+  , averageAge   :: Double
+  } deriving Show
+```
+
+* Add a notion of failure
+
+## The Data Model
+
+* Define our basic candidate type
+
+```haskell
+data Candidate = Candidate
+  { constituency :: String
+  , party        :: String
+  , sex          :: String
+  , averageAge   :: Double
+  } deriving Show
+```
+
+* Add a notion of failure
+
+```haskell
+type DataModel = Maybe Candidate
+```
+
+## The Data Model
+
+* Define our basic candidate type
+
+```haskell
+data Candidate = Candidate
+  { constituency :: String
+  , party        :: String
+  , sex          :: String
+  , averageAge   :: Double
+  } deriving Show
+```
+
+* Add a notion of failure
+
+```haskell
+type DataModel = Maybe Candidate
+```
+
+* Create a list of candidates
+
+## The Data Model
+
+* Define our basic candidate type
+
+```haskell
+data Candidate = Candidate
+  { constituency :: String
+  , party        :: String
+  , sex          :: String
+  , averageAge   :: Double
+  } deriving Show
+```
+
+* Add a notion of failure
+
+```haskell
+type DataModel = Maybe Candidate
+```
+
+* Create a list of candidates
+
+```haskell
+type DataModel = [Maybe Candidate]
+```
+
+
+
+
+# Data Transformation
+
+## Goal
+
+* Write some functions to:
+  * Convert CSV into data model
+  * Convert data model into JSON
```

Comment:
We're still introducing json at this step?

Reply:
Right at the end, yes